airbyte-integrations/connectors/source-google-ads/source_google_ads/streams.py

```diff
@@ -0,0 +1,144 @@
+    """
+    intervals = []
+    end_date = pendulum.parse(end_date) if end_date else pendulum.now()
+    start_date = pendulum.parse(start_date)
+
+    # As in to return some state when state in abnormal
+    if start_date > end_date:
+        return [{field: start_date.to_date_string()}]
+
+    # applying conversion window
+    start_date = start_date.subtract(days=conversion_window)
+
+    # Each stream_slice contains the beginning and ending timestamp for a 24 hour period
+    while start_date < end_date:
+        intervals.append({field: start_date.to_date_string()})
+        start_date = start_date.add(months=1)
+
+    return intervals
+
+
+class GoogleAdsStream(Stream, ABC):
+    def __init__(self, api: GoogleAds):
+        self.google_ads_client = api
+
+    def get_query(self, stream_slice: Mapping[str, Any]) -> str:
+        query = GoogleAds.convert_schema_into_query(schema=self.get_json_schema(), report_name=self.name)
+        return query
+
+    def parse_response(self, response: SearchPager) -> Iterable[Mapping]:
+        for result in response:
+            record = self.google_ads_client.parse_single_result(self.get_json_schema(), result)
+            validate(instance=record, schema=self.get_json_schema())
+            yield record
+
+    def read_records(self, sync_mode, stream_slice: Mapping[str, Any] = None, **kwargs) -> Iterable[Mapping[str, Any]]:
+        response = self.google_ads_client.send_request(self.get_query(stream_slice))
+        yield from self.parse_response(response)
+
+
+class IncrementalGoogleAdsStream(GoogleAdsStream, ABC):
+    def __init__(self, start_date: str, conversion_window_days: int, **kwargs):
+        self.conversion_window_days = conversion_window_days
+        self._start_date = start_date
+        super().__init__(**kwargs)
+
+    def stream_slices(self, stream_state: Mapping[str, Any] = None, **kwargs) -> Iterable[Optional[Mapping[str, any]]]:
+        stream_state = stream_state or {}
+        start_date = stream_state.get(self.cursor_field) or self._start_date
+
+        return chunk_date_range(start_date=start_date, conversion_window=self.conversion_window_days, field=self.cursor_field)
+
+    @staticmethod
+    def get_date_params(stream_slice: Mapping[str, Any], cursor_field: str, end_date: pendulum.datetime = None):
+        end_date = end_date or pendulum.yesterday()
+        start_date = pendulum.parse(stream_slice.get(cursor_field))
+        if start_date > pendulum.now():
+            return start_date.to_date_string(), start_date.add(days=1).to_date_string()
+
+        end_date = min(end_date, pendulum.parse(stream_slice.get(cursor_field)).add(months=1))
+        return start_date.add(days=1).to_date_string(), end_date.to_date_string()
+
+    def get_updated_state(self, current_stream_state: MutableMapping[str, Any], latest_record: Mapping[str, Any]) -> Mapping[str, Any]:
+        current_stream_state = current_stream_state or {}
```

Comment:
I saw that sometimes in some streams we modify `current_stream_state` dict in `get_updated_state()` function and sometimes we return new dict instance from `get_updated_state()` function. It's not consistent, I think. So 
why don't we create a new dict from `current_stream_state` and not modify `current_stream_state` itself for new connectors like this, for example:

```suggestion
        import copy  # move to import section
        current_stream_state = copy.deepcopy(current_stream_state) or {}
```

And maybe someday we will go through all connectors and do the same so all connectors will behave in the same way with `current_stream_state` and return new dict instances each time.